### PR TITLE
All entities can and are animated

### PR DESCRIPTION
**Pull request by Whixx**
Zombie: Attack, run.
Player: Ranged attack, melee attack, run, idle.
Spider: Attack (bite), run.
Demon: Walk, Idle, Attack (ranged).

### CODE REVIEW

File: StortSpelProjekt/Engine/src/ECS/Components/AnimationComponent.h [hunk 1]
```diff
@@ -18,16 +25,51 @@ namespace component
...
+        friend class ShadowRenderTask;
+        
+        AnimatedModel* m_pAnimatedModel = nullptr;
+        SkeletonNode* m_pSkeleton = nullptr;
+
+        DirectX::XMFLOAT4X4 m_GlobalInverseTransform;
+        std::map<std::string, TransformKey> m_AnimationState;
+
+        std::vector<Animation*> m_Animations;
+        std::pair<Animation*, AnimationInfo> m_pPendingAnimation;
```
Comment (fredrix96) [inline StortSpelProjekt/Engine/src/ECS/Components/AnimationComponent.h]: I am pretty sure you don't need a p in the names here

**Review by fredrix96 — CHANGES_REQUESTED** [review]
The demon's attack animation does not work as intended and I would really like the animations to stop when the enemies die

File: StortSpelProjekt/Engine/src/ECS/Components/AnimationComponent.h (hunk 1 above)
Comment (LeoWikstrom) [inline StortSpelProjekt/Engine/src/ECS/Components/AnimationComponent.h]: Not a pointer, should not have the p

File: StortSpelProjekt/Engine/src/Renderer/Animation.h [hunk 2]
```diff
@@ -18,10 +18,22 @@ struct TransformKey
...
 	DirectX::XMFLOAT3 position = {};
 	DirectX::XMFLOAT3* pPosition = &position;
-	DirectX::XMFLOAT4 rotationQuaternion = {};
-	DirectX::XMFLOAT4* pRotation = &rotationQuaternion;
+	DirectX::XMFLOAT4 rotation = {};
+	DirectX::XMFLOAT4* pRotation = &rotation;
 	DirectX::XMFLOAT3 scaling = { 1.0f,1.0f,1.0f };
 	DirectX::XMFLOAT3* pScale = &scaling;
+
+	//TransformKey& operator= (const TransformKey& other)
```
Comment (LeoWikstrom) [inline StortSpelProjekt/Engine/src/Renderer/Animation.h]: Remove comment if not needed

File: StortSpelProjekt/Engine/src/Renderer/Animation.h [hunk 3]
```diff
@@ -51,6 +63,14 @@ struct SkeletonNode
 	DirectX::XMFLOAT4X4 inverseBindPose;	// Bone offset
 	TransformKey* currentStateTransform;
 
+	SkeletonNode()
```
Comment (LeoWikstrom) [inline StortSpelProjekt/Engine/src/Renderer/Animation.h:54]: Why is this a struct and not a class?

File: StortSpelProjekt/Engine/src/Renderer/DX12Tasks/ForwardRenderTask.h [hunk 4]
```diff
@@ -3,16 +3,16 @@
 
 #include "RenderTask.h"
 
-class FowardRenderTask : public RenderTask
+class ForwardRenderTask : public RenderTask
```
Comment (LeoWikstrom) [inline StortSpelProjekt/Engine/src/Renderer/DX12Tasks/ForwardRenderTask.h:6]: 😞

**Review by LeoWikstrom — CHANGES_REQUESTED** [review]
A few requests

File: StortSpelProjekt/Game/src/Gamefiles/Components/AiComponent.cpp [hunk 5]
```diff
@@ -568,6 +612,12 @@ void component::AiComponent::updateRange(double dt)
 			{
 				if (m_IntervalTimeAccumulator >= m_AttackInterval && m_DistanceToPlayer >= m_AttackingDistance / 8.0f)
 				{
+					if (m_ShouldPlayAttackAnimation)
```
Comment (LeoWikstrom) [inline StortSpelProjekt/Game/src/Gamefiles/Components/AiComponent.cpp]: After testing, I found that the actual shot happens noticeably after the enemy moves his arm. I would suggest checking in this if-statement whether m_SpeedTimeAccumulator is larger than something like m_AttackSpeed / 2. You'll have to test and see which number works best, but something like that should give a more desirable result

**Review by LeoWikstrom — DISMISSED** [review]

**Review by fredrix96 — APPROVED** [review]
No memory leaks, demons attacks work, spiders animation work as well, although a little slow, no crash after a few rounds of playing. Life goes on, we are done!

**Review by OtakuFreak13 — APPROVED** [review]
Animations looks good enough for our game and quality standard.

**Pull request merged**